Add `CurrencyCode` type for the Currency `code` property

albarin opened the pull request:
Currently, the currency interface accepts any string as currency code, which would allow entering incorrect values.

This PR adds a new `type CurrencyCode` used as the type for the `code` and `currency_code` properties. That way the possible string values are restricted only to the allowed _ISO 4217_ codes.

Fixes https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5397

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.

### Manual Testing
1. Create a new page with the "All Products" and the "Filter Products by Price" blocks.
2. Go to the page on the front
3. Check that the price/currency is showing ok on the slider inputs and the product prices.

-

1. Add a product to the cart
2. Go to the checkout
3. Check the price and currency are ok

albarin committed:
Create the CurrencyCode and use it for the currency `code` instead of the type `string`

Using the CurrencyCode type, the allowed values for the `code` property are restricted
to the ISO 4217 codes and not any string.

tomasztunik commented on packages/prices/utils/price.ts:
```diff
@@ -90,7 +90,7 @@ export const getCurrencyFromPriceResponse = (
 	} = currencyData;
 
-	return {
+	return <Currency>{
```
Casting is not required here, it will validate returned object against the return type provided in function definition at line 78

albarin replied:
Oh okay, the IDE was complaining about it 🤔

albarin committed:
Remove unnecessary casting

albarin committed:
Add and remove some currency codes to match the ones on WooCommerce

The WooCommerce supported list can be found here: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-core-functions.php#L475

albarin committed:
Use CurrencyCode type on CurrencyInfo and CurrencyResponse instead of string

sunyatasattva commented:
Link to the Slack thread where it was decided to go for union types on this: p1646913422664749-slack-C02SGH7JBGS

sunyatasattva commented:
I've added the label `skip-changelog` and removed the changelog entry. This is an internal change and should probably not be listed in the changelog which is exposed to our users.

sunyatasattva reviewed (approved):
I also checked that WooCommerce has uppercase Currency Codes, so that's nice. I'll approve this and also merging it in for you by skipping the required CI steps. The failed E2E is due to something unrelated to this and to the newest Gutenberg version causing a regression. A fix might be available here: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6080

We made sure the E2E tests would otherwise pass.

However, I'd probably amend the manual testing steps in the PR description itself, to make sure to test a few more components which consume the `Currency` type.

So, nothing code related.
